vars/copyContainer.groovy

```diff
@@ -10,34 +10,67 @@
+    source_image_no_tag = source_image.split(':')[0]
+    source_registry = args.sourceRegistry
+    destination_image = args.destinationImage
+    destination_image_no_tag = destination_image.split(':')[0]
+    destination_registry = args.destinationRegistry
 
     if (args.destinationRegistry == 'opensearchstaging' || args.destinationRegistry == 'opensearchproject') {
         def dockerJenkinsCredential = args.destinationRegistry == 'opensearchproject' ? "jenkins-production-dockerhub-credential" : "jenkins-staging-dockerhub-credential"
         withCredentials([usernamePassword(credentialsId: dockerJenkinsCredential, usernameVariable: 'DOCKER_USERNAME', passwordVariable: 'DOCKER_PASSWORD')]) {
             def dockerLogin = sh(returnStdout: true, script: "set +x && echo $DOCKER_PASSWORD | docker login --username $DOCKER_USERNAME --password-stdin").trim()
-            sh """gcrane cp ${args.sourceRegistry}/${args.sourceImage} ${args.destinationRegistry}/${args.destinationImage}; docker logout"""
+            gcraneCopy()
         }
     }
+
     if (args.destinationRegistry == 'public.ecr.aws/opensearchproject') {
         withCredentials([
             string(credentialsId: 'jenkins-artifact-promotion-role', variable: 'ARTIFACT_PROMOTION_ROLE_NAME'),
             string(credentialsId: 'jenkins-aws-production-account', variable: 'AWS_ACCOUNT_ARTIFACT')])
             {
                 withAWS(role: "${ARTIFACT_PROMOTION_ROLE_NAME}", roleAccount: "${AWS_ACCOUNT_ARTIFACT}", duration: 900, roleSessionName: 'jenkins-session') {
                     def ecrLogin = sh(returnStdout: true, script: "set +x && aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin ${args.destinationRegistry}").trim()
-                    sh """gcrane cp ${args.sourceRegistry}/${args.sourceImage} ${args.destinationRegistry}/${args.destinationImage}; docker logout ${args.destinationRegistry}"""
+                    gcraneCopy()
                 }
             }
     }
+
     if(args.destinationRegistry == 'public.ecr.aws/opensearchstaging') {
-            def ecrLogin = sh(returnStdout: true, script: "set +x && aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin ${args.destinationRegistry}").trim()
-            sh """gcrane cp ${args.sourceRegistry}/${args.sourceImage} ${args.destinationRegistry}/${args.destinationImage}; docker logout ${args.destinationRegistry}"""
+        def ecrLogin = sh(returnStdout: true, script: "set +x && aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin ${args.destinationRegistry}").trim()
+        gcraneCopy()
     }
+
+}
+
+def gcraneCopy() {
+    sh """
+        set +x
+
+        if [ ${recursive_copy} = true ]; then
+            echo "Copying all image tags recursively from ${source_registry}/${source_image_no_tag} to ${destination_registry}/${destination_image_no_tag}"
```

Comment:
If I understand correctly at any given moment the recursive copy will only happen between values provided in `source_registry` and `destination_registry`. Meaning if the values are `opensearchproject` and `opensearchstaging` then all the tags for a given image will get copied from `opensearchproject` to `opensearchstaging` but no copy will happen to ecr repo. Right?

Reply:
I think my point is very much similar to what @gaiksaya is asking.

Reply:
If you specify source_reg as ecrstaging, and dest_reg as ecrprod, yes it will copy all the tags of a given image from ecrstaging to ecrprod, vise versa. If you specify another registry like gcp container registry it is the same thing, it will copy from one registry to another registry, copy all the tags of a image between given registries of the image.

Also, yeah it is isolated only between the given registries, recursively copy all the tags of a given image, not iterate among all the registries at all.

Reply:
This is really risky and with very high blast radius. I don't want anyone mistakenly copying all tags from opensearchstaging to opensearchProject dockerhub. 
One mis-click and all images will be overwritten. Why would we want to copy all tags from one source to destination? One copy at a time is good. I believe we should not add recursive copy instead call this method recursively in jenkins file if required. 

cc: @bbarani @prudhvigodithi Needs your input here.

Reply:
If needed I can move this to its own workflow.
And I think it is very beneficial.

We can have a workflow running on cron to sync images between hub and ecr such like ci-runner images, and we can have a quick sync without human intervention. 

This can also be a active backup function so we dont have to manually do multiple copies just to create one image and forget to backup in ECR or other registry.

We can use the workflow jenkinsfile to lock down source and destination, and this can happen between only staging and staging.

All of our ci images are versioned, means the risk is not as high. All in all, this is to provide a mean for us to easily manage our sync images and backup images.

Have a workflow slowly schedule one to one copy can be very expansive and time consuming, especially if it is for backup images. And we already have a lot of tags right now. It is not going to scale on the long run.

Thanks.

Reply:
In that case I believe we should only allow recursive for CI-runner images where source and destination can either be `opensearchstaging` and/or `public.ecr.aws/opensearchstaging` and repo name being ci-runner.
The same environment also contains credentials for prod and other opensearch images which we do not want to sync for all tags anytime between staging and Prod for any version. 

Also can you add tests for default and non-default behavior?
Thanks!

Reply:
We already have tests for default vs non-default behavior, will add restrictions of destination and repo name.

Reply:
Added now.